[Flaky test] TestPodLabelsUpdate/Update_Pod_labels_to_match_TrafficControl_tc1_(effective),_tc3_(alternative)

**Describe the bug**
I have seen this test fail in CI.

```
2022-09-02T00:29:19.8761094Z --- FAIL: TestPodLabelsUpdate (0.98s)
2022-09-02T00:29:19.8762157Z     --- FAIL: TestPodLabelsUpdate/Update_Pod_labels_to_match_TrafficControl_tc1_(effective),_tc3_(alternative) (0.12s)
2022-09-02T00:29:19.8762789Z         controller_test.go:929: 
2022-09-02T00:29:19.8763710Z             	Error Trace:	/home/runner/work/antrea/antrea/pkg/agent/controller/trafficcontrol/controller_test.go:929
2022-09-02T00:29:19.8764238Z             	Error:      	Not equal: 
2022-09-02T00:29:19.8765872Z             	            	expected: &trafficcontrol.podToTCBinding{effectiveTC:"test-tc1", alternativeTCs:sets.String{"test-tc3":sets.Empty{}}}
2022-09-02T00:29:19.8767856Z             	            	actual  : &trafficcontrol.podToTCBinding{effectiveTC:"test-tc1", alternativeTCs:sets.String{"test-tc2":sets.Empty{}, "test-tc3":sets.Empty{}}}
2022-09-02T00:29:19.8768403Z             	            	
2022-09-02T00:29:19.8768800Z             	            	Diff:
2022-09-02T00:29:19.8769345Z             	            	--- Expected
2022-09-02T00:29:19.8770956Z             	            	+++ Actual
2022-09-02T00:29:19.8771587Z             	            	@@ -2,3 +2,5 @@
2022-09-02T00:29:19.8772463Z             	            	  effectiveTC: (string) (len=8) "test-tc1",
2022-09-02T00:29:19.8773329Z             	            	- alternativeTCs: (sets.String) (len=1) {
2022-09-02T00:29:19.8774096Z             	            	+ alternativeTCs: (sets.String) (len=2) {
2022-09-02T00:29:19.8775003Z             	            	+  (string) (len=8) "test-tc2": (sets.Empty) {
2022-09-02T00:29:19.8775438Z             	            	+  },
2022-09-02T00:29:19.8776318Z             	            	   (string) (len=8) "test-tc3": (sets.Empty) {
2022-09-02T00:29:19.8777231Z             	Test:       	TestPodLabelsUpdate/Update_Pod_labels_to_match_TrafficControl_tc1_(effective),_tc3_(alternative)
2022-09-02T00:29:19.8777599Z FAIL
2022-09-02T00:29:19.8778095Z coverage: 2.2% of statements in antrea.io/antrea/cmd/..., antrea.io/antrea/pkg/..., antrea.io/antrea/multicluster/cmd/..., antrea.io/antrea/multicluster/controllers/...
2022-09-02T00:29:19.8778653Z FAIL	antrea.io/antrea/pkg/agent/controller/trafficcontrol	11.302s
```

--- a/pkg/agent/controller/trafficcontrol/controller_test.go
+++ b/pkg/agent/controller/trafficcontrol/controller_test.go
@@ -62,6 +62,15 @@ type fakeController struct {
 	podUpdateChannel    *channel.SubscribableChannel
 }
 
+func (c *fakeController) startInformers(stopCh chan struct{}) {
+	c.informerFactory.Start(stopCh)
+	c.informerFactory.WaitForCacheSync(stopCh)
+	go c.localPodInformer.Run(stopCh)
+	cache.WaitForCacheSync(stopCh, c.localPodInformer.HasSynced)
+	c.crdInformerFactory.Start(stopCh)
+	c.crdInformerFactory.WaitForCacheSync(stopCh)
+}
+
 var (
 	labels1 = map[string]string{"app1": "foo1"}
 	labels2 = map[string]string{"app2": "foo2"}
@@ -475,11 +484,7 @@ func TestTrafficControlAdd(t *testing.T) {
 			stopCh := make(chan struct{})
 			defer close(stopCh)
 
-			c.informerFactory.Start(stopCh)
-			c.informerFactory.WaitForCacheSync(stopCh)
-			go c.localPodInformer.Run(stopCh)
-			c.crdInformerFactory.Start(stopCh)
-			c.crdInformerFactory.WaitForCacheSync(stopCh)
+			c.startInformers(stopCh)
 
 			tt.expectedCalls(c.mockOFClient, c.mockOVSBridgeClient, c.mockOVSCtlClient)
 			assert.NoError(t, c.syncTrafficControl(tt.tc.Name))
@@ -572,11 +577,7 @@ func TestTrafficControlUpdate(t *testing.T) {
 			stopCh := make(chan struct{})
 			defer close(stopCh)
 
-			c.informerFactory.Start(stopCh)
-			c.informerFactory.WaitForCacheSync(stopCh)
-			go c.localPodInformer.Run(stopCh)
-			c.crdInformerFactory.Start(stopCh)
-			c.crdInformerFactory.WaitForCacheSync(stopCh)
+			c.startInformers(stopCh)
 
 			// Fake the status after TrafficControl tc1 is added.
 			c.portToTCBindings = map[string]*portToTCBinding{
@@ -602,13 +603,14 @@ func TestTrafficControlUpdate(t *testing.T) {
 			item, _ := c.queue.Get()
 			c.queue.Done(item)
 
+			tt.updatedTrafficControl.Generation += 1
 			_, err := c.crdClient.CrdV1alpha2().TrafficControls().Update(context.TODO(), tt.updatedTrafficControl, metav1.UpdateOptions{})
 			require.NoError(t, err)
 
 			// Functions are expected to be called after updating TrafficControl tc1.
 			tt.expectedCalls(c.mockOFClient, c.mockOVSBridgeClient, c.mockOVSCtlClient)
 
-			time.Sleep(time.Second)
+			waitEvents(t, 1, c)
 			require.NoError(t, c.syncTrafficControl(tc1Name))
 			require.Equal(t, tt.expectedState, c.tcStates[tc1Name])
 		})
@@ -631,11 +633,7 @@ func TestSharedTargetPort(t *testing.T) {
 	stopCh := make(chan struct{})
 	defer close(stopCh)
 
-	c.informerFactory.Start(stopCh)
-	c.informerFactory.WaitForCacheSync(stopCh)
-	go c.localPodInformer.Run(stopCh)
-	c.crdInformerFactory.Start(stopCh)
-	c.crdInformerFactory.WaitForCacheSync(stopCh)
+	c.startInformers(stopCh)
 
 	// Target port is expected to be crated if it doesn't exist.
 	c.mockOVSBridgeClient.EXPECT().CreatePort(targetPort1Name, targetPort1Name, externalIDs)
@@ -688,11 +686,7 @@ func TestPodUpdateFromCNIServer(t *testing.T) {
 	stopCh := make(chan struct{})
 	defer close(stopCh)
 
-	c.informerFactory.Start(stopCh)
-	c.informerFactory.WaitForCacheSync(stopCh)
-	go c.localPodInformer.Run(stopCh)
-	c.crdInformerFactory.Start(stopCh)
-	c.crdInformerFactory.WaitForCacheSync(stopCh)
+	c.startInformers(stopCh)
 	go c.podUpdateChannel.Run(stopCh)
 
 	// Fake the status after TrafficControl tc1 is added.
@@ -833,11 +827,7 @@ func TestPodLabelsUpdate(t *testing.T) {
 			stopCh := make(chan struct{})
 			defer close(stopCh)
 
-			c.informerFactory.Start(stopCh)
-			c.informerFactory.WaitForCacheSync(stopCh)
-			go c.localPodInformer.Run(stopCh)
-			c.crdInformerFactory.Start(stopCh)
-			c.crdInformerFactory.WaitForCacheSync(stopCh)
+			c.startInformers(stopCh)
 
 			// Fake the status after TrafficControl tc1, tc2 and tc3 is added. TrafficControl tc1 is the effective
 			// TrafficControl of the Pod, and tc2 is the alternative TrafficControl of the Pod.
@@ -1014,11 +1004,7 @@ func TestNamespaceLabelsUpdate(t *testing.T) {
 			stopCh := make(chan struct{})
 			defer close(stopCh)
 
-			c.informerFactory.Start(stopCh)
-			c.informerFactory.WaitForCacheSync(stopCh)
-			go c.localPodInformer.Run(stopCh)
-			c.crdInformerFactory.Start(stopCh)
-			c.crdInformerFactory.WaitForCacheSync(stopCh)
+			c.startInformers(stopCh)
 
 			// Fake the status after TrafficControl tc1, tc2 and tc3 is added. TrafficControl tc1 is the effective
 			// TrafficControl of the Pod, and tc2 is the alternative TrafficControl of the Pod.
@@ -1130,11 +1116,7 @@ func TestPodDelete(t *testing.T) {
 	stopCh := make(chan struct{})
 	defer close(stopCh)
 
-	c.informerFactory.Start(stopCh)
-	c.informerFactory.WaitForCacheSync(stopCh)
-	go c.localPodInformer.Run(stopCh)
-	c.crdInformerFactory.Start(stopCh)
-	c.crdInformerFactory.WaitForCacheSync(stopCh)
+	c.startInformers(stopCh)
 
 	// Fake the status after TrafficControl tc1, tc2 and tc3 is added. TrafficControl tc1 is the effective
 	// TrafficControl of the Pod, and tc2, tc3 is the alternative TrafficControl of the Pods.

